docs/adr/0003-recording-service-usage-in-account.md

```diff
@@ -0,0 +1,98 @@
+
+This will then be read by a series of GOV.UK Account team managed Lambdas that process and store which services each user has used.
+
+Code standards follow [GOV.UK One Login's Serverless Best Practices](https://govukverify.atlassian.net/wiki/spaces/DISBP/overview) for lambdas and [DI Dev Platform](https://govukverify.atlassian.net/wiki/spaces/PLAT/overview) guides for deployment.
+
+Code and configuration will be stored in a new repository [di-account-management-backend](https://github.com/alphagov/di-account-management-backend).
+
+## Context
+
+There is the requirement to allow users to see which services they have accessed, and have that data stored against their account, so we can personalise the account management user interface.
+
+See also [alphagov/digital-identity-architecture#279](https://github.com/alphagov/digital-identity-architecture/pull/279) for context on data sources.
+
+We will constrain ourselves by prior ADRs that limits options here.
+
+[ADR 0038 Backend Node.js/TypeScript](https://github.com/alphagov/digital-identity-architecture/blob/main/adr/0038-backend-nodejs-typescript.md) and [ADR 0037 Using external dependencies in production with Node.js/NPM](https://github.com/alphagov/digital-identity-architecture/blob/main/adr/0037-nodejs-npm-production-dependencies.md) guide the choice of Node as a default, we have no cause to deviate.
+
+We have reviewed [ADR 0025 Relaxing serverless constraints for MVP](https://github.com/alphagov/digital-identity-architecture/blob/055042de6374a99d646272022177a0d8da6fcd4e/adr/0025-relaxing-serverless-constraints.md) and believe that we can provide a serverless implimentation within our delivery window that meets [GOV.UK One Login's Serverless Best Pratices](https://govukverify.atlassian.net/wiki/spaces/DISBP/overview) and [DI Dev Platform deployment standards](https://govukverify.atlassian.net/wiki/spaces/PLAT/overview), so we have chosen to avoid accruing further technical debt.
+
+These previous decisions focus us towards the option below.
+
+A key assumption is that standard AWS security practices of access controls and default DynamoDB encryption meet our security standards given the sensitivity of user data stored. Data will be encrypted at rest with an AWS managed key, as [the default encryption configuration](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/EncryptionAtRest.html) for a DynamoDB table
+
+## Decision
+
+As a result of prior ADRs we have decided to:
+
+- Use DynamoDB as our data store, encrypted at rest with a non-default KMS key
+- Write small, single responsibility lambdas written in Typescript with minimal production dependencies
+- Separate of lambdas with SQS Queues to allow for async processing and scaling
+- Use dead letter queues to monitor and capture data from handled errors.
+
```

Comment:
```suggestion

We will store a single record per user and update that record as the users logs in to more services. This will make querying for a user's data very fast which is important, as the query will be blocking the synchronous user-facing HTTP request to load the service dashboard.

This means the time between us receiving an event from TxMA and writing it to the data store will be slightly longer since we're doing the data processing upfront. That's ok because data ingestion is an asynchronous, non-user facing process. Based on testing we expect it to take under 1 second in the worst case scenario when each lambda has to cold start which is an acceptable latency.

```
I think we should add this section to explain why we chose this specific architecture. We could have gone with a model that stores the incoming events as is (and actually I think we should still store them - will raise a separate ADR to discuss) but we decided not to because that would increase query time in the user facing part of the system.